Document how to connect via ssh (#501)
nebari-dev/nebari-docs: docs/docs/how-tos/connect-via-ssh.mdx

---
id: connect-via-ssh
title: Connect to Nebari via SSH or SFTP
description: Set up a remote SSH connection using jupyter-ssh
---

# Connect to Nebari via SSH or SFTP

> Secure Shell Protocol (SSH) is a common protocol for communicating securely over a network connection. Secure File Transfer Protocol (SFTP) is based on SSH and is used to transfer files to and from a remote machine. In this guide, you will learn how to connect to your Nebari deployment via SSH, and how to transfer files using SFTP from your local machine. 

Reviewer: ```suggestion
Secure Shell Protocol (SSH) is a standard protocol for communicating securely over a network connection. Secure File Transfer Protocol (SFTP) is based on SSH and transfers files to and from a remote machine. In this guide, you will learn how to connect and launch your JupyterLab user instance via SSH and transfer files using SFTP from your local machine. 
```

> Nebari includes a package called [`jupyterhub-ssh`](https://github.com/yuvipanda/jupyterhub-ssh) in all deployments to allow remote connection to Nebari deployments. On Nebari, SSH connections are set up on port 8022 and SFTP connections on port 8023. 

Reviewer: ```suggestion
Nebari includes a package called [`jupyterhub-ssh`](https://github.com/yuvipanda/jupyterhub-ssh) in all deployments to allow remote connection to their user instances. On Nebari, SSH connections are set up on port 8022 and SFTP connections on port 8023. 
```

## Prepare for the connection on the Nebari deployment

> Nebari, at its core, is running on a JupyterHub server. You will create a JupyterHub _token_ in order to access your server remotely. To do this, log into your Nebari deployment and navigate to the JupyterHub token page, {your-nebari-domain.org/hub/token}. 

Reviewer 2: It'd be nice to include how to navigate via the GUI - 
![image](https://github.com/user-attachments/assets/22c0df86-231e-4dc0-ba00-20f97c3443c5)

Reviewer 2: The curly brackets seem strange to me.  I would just use normal parenthesis.

> Nebari, at its core, is running on a JupyterHub server. You will create a JupyterHub _token_ in order to access your server remotely. To do this, log into your Nebari deployment and navigate to the JupyterHub token page, {your-nebari-domain.org/hub/token}. 

Author: ```suggestion
Nebari, at its core, is running on a JupyterHub server. You will create a JupyterHub _token_ in order to access your server remotely. To do this, log into your Nebari deployment and navigate to the JupyterHub token page, \<your-nebari-domain.org/hub/token\>. 
```

> Nebari, at its core, is running on a JupyterHub server. You will create a JupyterHub _token_ in order to access your server remotely. To do this, log into your Nebari deployment and navigate to the JupyterHub token page, \<your-nebari-domain.org/hub/token\>. 
> 

Reviewer: uhmmm... this is not true, though. A jupyterhub server runs on nebari, not the other way around.
Instead I would just remove and let those be part of the previous paragraph
```suggestion
You will create a JupyterHub _token_ in order to access your server remotely. To do this, log into your Nebari deployment and navigate to the JupyterHub token page, \<your-nebari-domain.org/hub/token\>. 

```

This page is also accessible via the Nebari Home Page:

* If your deployment has `jhub-apps` enabled, you can navigate to this page by selecting the `Token` option under your username in the upper righthand corner. 
* If your deployment does not have `jhub-apps` enabled, you can click on the `Token` link on the top navigation bar. 

You may add a note and expiration date to the token. You can leave the scopes field blank to give the token all the permissions that your user typically has on JupyterHub (which is the service on which Nebari is running).

> ## Connect to Nebari via SSH

Reviewer: this title needs to be changed, same reason as previous

Before you being, you'll need to have SSH installed on your local machine. From a terminal, use `ssh` to begin the connection. 

> `ssh {your-nebari-domain.com} -l {your-username} -p 8022`

Reviewer 2: We've typically used angle brackets instead of curly brackets for usage like this. E.g.
![image](https://github.com/user-attachments/assets/77bcb9ad-bbd2-48b3-9373-2dcee94e0d01)

Actually, I seem to see inconsistent usage of angle and curly brackets
![image](https://github.com/user-attachments/assets/a1a435f9-b97a-4f0d-a923-9b8fef72e85b)

Author: Yeah, our existing docs are inconsistent. I can switch to the angle brackets instead though. I *think* there might be more of those (after seeing so many of them throw errors in my docusaurus v3 PR...).

Reviewer 2: Could you open an issue to remind us to resolve the curly brackets/angle bracket inconsistency at some point?

Author: https://github.com/nebari-dev/nebari-docs/issues/510

> `ssh {your-nebari-domain.com} -l {your-username} -p 8022`

Author: ```suggestion
`ssh \<your-nebari-domain.com\> -l \<your-username\> -p 8022`
```

:::note

> The syntax for `{your-nebari-domain.com}` is important - do not include `https://` or a trailing `/`. 

Author: ```suggestion
The syntax for `\<your-nebari-domain.com\>` is important - do not include `https://` or a trailing `/`. 
```

> If you have the incorrect syntax, it will result in the error: `Could not resolve hostname {your-nebari-domain.com}: nodename nor servname provided, or not known`

Author: ```suggestion
If you have the incorrect syntax, it will result in the error: `Could not resolve hostname \<your-nebari-domain.com\>: nodename nor servname provided, or not known`
```

Nebari itself is generally accessed via [kubectl](/docs/how-tos/debug-nebari#installing-kubectl)
to connect directly to the cluster (nodes or pods), while the JupyterHub server runs _on_ Nebari and controls the spawning of
user instances of JupyterLab. The approach described here uses a JupyterHub plugin, `jupyterhub-ssh`, to spawn and access users' JupyterLab instance
:::

## Prepare for the connection on the Nebari deployment

You will create a JupyterHub _token_ in order to access your server remotely. To do this, log into your Nebari deployment and navigate to the JupyterHub token page, <your-nebari-domain.org/hub/token>.

This page is also accessible via the Nebari Home Page:

- If your deployment has `jhub-apps` enabled, you can navigate to this page by selecting the `Token` option under your username in the upper righthand corner.
- If your deployment does not have `jhub-apps` enabled, you can click on the `Token` link on the top navigation bar.

You may add a note and expiration date to the token. You can leave the scopes field blank to give the token all the permissions that your user typically has on JupyterHub (which is the service on which Nebari is running).

> 

Author: added!

`Starting your server............................................failed to start server on time!`

The good news is that Nebari is still working on spinning up your server. Retry the `ssh` connection until the connection is established once the server is up and accessible.

Once the connection is complete, you will have terminal access to your JupyterLab instance on Nebari.

## Use SFTP to transfer files to/from Nebari

SFTP, which stands for _Secure File Transfer Protocol_, is a protocol built into SSH that can implement file transfer commands over a secure connection. The SFTP port on Nebari is 8023.

Utilizing SFTP will depend on how your local machine is configured and the utility you're using for SFTP. In some cases, your SSH configuration will allow SFTP to work "out of the box". Other times, special SFTP setup is required in your SSH configuration to allow for these connections.

> 

Reviewer: ```suggestion
Similar to `ssh`, from a terminal, use `sftp` to begin a new connection:
```sftp <your-nebari-domain.com> -l <your-username> -p 8023```

### Transferring Files with SFTP

Downloading Files:

- Use the `get` command to download files from the remote host:
	- `get <remote-file>` downloads the file with the same name to the local system.
	- `get <remote-file> <local-file>` downloads and saves the file with a different name.
	- To download a directory and its contents, use `get -r <some-directory>`.
	- To maintain permissions and access times, use `get -Pr <some-directory>`.

Uploading Files:

- Use the `put` command to upload files to the remote system:
	- `put <local-file>` uploads the file.
	- To upload a directory and its contents, use `put -r <some-directory>`.
```

`Starting your server............................................failed to start server on time!`

The good news is that Nebari is still working on spinning up your server. Retry the `ssh` connection until the connection is established once the server is up and accessible.

Once the connection is complete, you will have terminal access to your JupyterLab instance on Nebari.

## Use SFTP to transfer files to/from Nebari

By default, SFTP uses the SSH protocol to authenticate and establish a secure connection. Because of this, the same authentication token used previously in its SSH counterpart can be used here.

Utilizing SFTP will depend on how your local machine is configured and the utility you're using for SFTP. In some cases, your SSH configuration will allow SFTP to work "out of the box". Other times, special SFTP setup is required in your SSH configuration to allow for these connections.
Similar to `ssh`, from a terminal, use `sftp` to begin a new connection:

> `sftp <your-nebari-domain.com> -l <your-username> -p 8023`

Reviewer: ```suggestion

`sftp <your-nebari-domain.com> -l <your-username> -p 8023`
```
